compiler/rustc_hir_analysis/src/impl_wf_check.rs

```diff
@@ -88,14 +88,36 @@ fn enforce_impl_params_are_constrained(
 
     impl_trait_ref.error_reported()?;
 
-    let mut input_parameters = cgp::parameters_for_impl(tcx, impl_self_ty, impl_trait_ref);
+    let mut constrained_parameters = cgp::parameters_for_impl(tcx, impl_self_ty, impl_trait_ref);
     cgp::identify_constrained_generic_params(
         tcx,
         impl_predicates,
         impl_trait_ref,
-        &mut input_parameters,
+        &mut constrained_parameters,
     );
 
+    // Reasons each generic is unconstrained, or Other if not present
+    let mut unconstrained_reasons = FxHashMap::default();
+    for (clause, _) in impl_predicates.predicates {
+        if let Some(projection) = clause.as_projection_clause() {
+            let mentioned_params = cgp::parameters_for(tcx, projection.term().skip_binder(), true);
+            let is_clause_circular =
+                Some(projection.required_poly_trait_ref(tcx).skip_binder()) == impl_trait_ref;
```

Comment:
Almost forgot, but wanted to add a quick note/question regarding `skip_binder` here... from what I understand, binders correspond to things like `for<'a>`, and `skip_binder` is *generally* a bad idea since it just unwraps the binder wo/ updating references to the quantifier.

My Q: is this specific usage safe? Or is some else more appropriate? (My understanding of this is a bit hazy 🙂)